File: services/ocr_service.py
"""
OCR service for extracting text and structured data from delivery dockets
Uses Tesseract OCR with preprocessing
"""

import pytesseract
import re
from typing import Dict, List, Optional
from datetime import datetime
import logging
from pathlib import Path
import config
from services.image_processor import ImageProcessor

logger = logging.getLogger(__name__)

# Configure Tesseract path
if config.TESSERACT_CMD:
    pytesseract.pytesseract.tesseract_cmd = config.TESSERACT_CMD


class OCRService:
    """Extract text and structured data from delivery dockets"""

    @staticmethod
    def extract_text(image_path: str, preprocess: bool = True) -> Dict:
        """
        Extract raw text from image using OCR
        Returns: {
            'text': str,
            'confidence': float
        }
        """
        try:
            if preprocess:
                # Preprocess image
                processed_img = ImageProcessor.preprocess_for_ocr(image_path)
            else:
                processed_img = ImageProcessor.load_image(image_path)

            # Run OCR with detailed output
            ocr_data = pytesseract.image_to_data(
                processed_img,
                output_type=pytesseract.Output.DICT,
                config='--psm 6'  # Assume uniform text block
            )

            # Extract text
            text = pytesseract.image_to_string(processed_img, config='--psm 6')

            # Calculate average confidence
            confidences = [int(conf) for conf in ocr_data['conf'] if int(conf) > 0]
            avg_confidence = sum(confidences) / len(confidences) if confidences else 0

            logger.info(f"OCR complete. Confidence: {avg_confidence:.1f}%")

            return {
                'text': text.strip(),
                'confidence': round(avg_confidence, 2),
                'word_count': len(text.split())
            }

        except Exception as e:
            logger.error(f"OCR error: {e}")
            raise

    @staticmethod
    def extract_docket_data(image_path: str) -> Dict:
        """
        Extract structured data from delivery docket
        Returns: {
            'supplier_name': str,
            'docket_number': str,
            'po_reference': str,
            'delivery_date': date,
            'line_items': List[Dict],
            'raw_text': str,
            'confidence': float
        }
        """
        # Get raw OCR text
        ocr_result = OCRService.extract_text(image_path)
        text = ocr_result['text']
        lines = text.split('\n')

        # Initialize extraction result
        result = {
            'supplier_name': None,
            'docket_number': None,
            'po_reference': None,
            'delivery_date': None,
            'line_items': [],
            'raw_text': text,
            'confidence': ocr_result['confidence']
        }

        # Extract supplier name (usually near top)
        result['supplier_name'] = OCRService._extract_supplier(lines)

        # Extract docket number
        result['docket_number'] = OCRService._extract_docket_number(text)

        # Extract PO reference
        result['po_reference'] = OCRService._extract_po_reference(text)

        # Extract delivery date
        result['delivery_date'] = OCRService._extract_date(text)

        # Extract line items
        result['line_items'] = OCRService._extract_line_items(lines)

        logger.info(f"Extracted: PO={result['po_reference']}, Docket={result['docket_number']}, Lines={len(result['line_items'])}")

        return result

    @staticmethod
    def _extract_supplier(lines: List[str]) -> Optional[str]:
        """Extract supplier name from top of document"""
        # Look in first 10 lines for company name indicators
        for line in lines[:10]:
            line = line.strip()
            # Skip empty lines and lines with common header words
            if not line or line.lower() in ['tax invoice', 'delivery docket', 'invoice', 'receipt']:
                continue
            # First substantial line is often company name
            if len(line) > 5 and not re.match(r'^[\d\s\-/]+$', line):
                return line
        return None

    @staticmethod
    def _extract_docket_number(text: str) -> Optional[str]:
        """Extract delivery docket number"""
        patterns = [
            r'(?:docket|delivery)\s*(?:no|number|#)[\s:]*([A-Z0-9\-]+)',
            r'(?:doc|dkt)\s*(?:no|#)[\s:]*([A-Z0-9\-]+)',
            r'delivery\s+([A-Z0-9]{5,})',
        ]

        for pattern in patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                return match.group(1).strip()

        return None

    @staticmethod
    def _extract_po_reference(text: str) -> Optional[str]:
        """Extract PO reference number"""
        patterns = [
            r'(?:PO|P\.O\.|Purchase Order)[\s#:]*([A-Z0-9\-]+[ABC]?)',
            r'(?:Order|Ref|Reference)[\s#:]*([A-Z0-9\-]{5,}[ABC]?)',
            r'\b(PO[-\s]?\d{4,}[ABC]?)\b',
        ]

        for pattern in patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                po_ref = match.group(1).strip().upper()
                # Clean up common OCR errors
                po_ref = po_ref.replace(' ', '').replace('.', '-')
                return po_ref

        return None

    @staticmethod
    def _extract_date(text: str) -> Optional[datetime]:
        """Extract delivery date"""
        date_patterns = [
            r'\b(\d{1,2}[/-]\d{1,2}[/-]\d{2,4})\b',
            r'\b(\d{4}[/-]\d{1,2}[/-]\d{1,2})\b',
            r'\b(\d{1,2}\s+(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+\d{2,4})\b',
        ]

        for pattern in date_patterns:
            matches = re.findall(pattern, text, re.IGNORECASE)
            for date_str in matches:
                try:
                    # Try various date formats
                    for fmt in ['%d/%m/%Y', '%d-%m-%Y', '%Y/%m/%d', '%Y-%m-%d', '%d/%m/%y', '%d %b %Y', '%d %B %Y']:
                        try:
                            return datetime.strptime(date_str, fmt).date()
                        except ValueError:
                            continue
                except Exception:
                    continue

        return None

    @staticmethod
    def _extract_line_items(lines: List[str]) -> List[Dict]:
        """
        Extract line items from docket
        Each line item contains: SKU, description, quantity
        """
        line_items = []
        in_items_section = False

        # Patterns for line items
        qty_pattern = r'\b(\d+(?:\.\d+)?)\b'
        sku_pattern = r'\b([A-Z0-9]{3,}[-_]?[A-Z0-9]*)\b'

        for i, line in enumerate(lines):
            line = line.strip()
            if not line:
                continue

            # Detect start of items section
            if re.search(r'(?:item|product|description|qty|quantity)', line, re.IGNORECASE):
                in_items_section = True
                continue

            # Detect end of items section
            if re.search(r'(?:total|subtotal|tax|gst|notes|signature)', line, re.IGNORECASE):
                break

            if in_items_section:
                # Extract quantity (usually a number)
                qty_matches = re.findall(qty_pattern, line)

                # Extract SKU (alphanumeric code)
                sku_matches = re.findall(sku_pattern, line)

                # Extract description (remaining text)
                desc = line

                if qty_matches:
                    line_items.append({
                        'line_number': len(line_items) + 1,
                        'sku': sku_matches[0] if sku_matches else None,
                        'description': desc,
                        'quantity': float(qty_matches[-1]),  # Last number is usually qty
                        'confidence': 70  # Default confidence
                    })

        logger.info(f"Extracted {len(line_items)} line items")
        return line_items

File: backend/src/services/ocrService.ts
import { createWorker, Worker } from 'tesseract.js';
import { imagePreprocessor } from './imagePreprocessor';
// @ts-ignore - pdf-parse doesn't have type definitions
import pdfParse from 'pdf-parse';
import { readFileSync } from 'fs';

export interface ExtractedDocketData {
  supplierName?: string;
  docketNumber?: string;
  poReference?: string;
  deliveryDate?: string;
  lineItems: Array<{
    sku?: string;
    description: string;
    quantity: number;
    unit?: string;
  }>;
  rawText: string;
  confidence: number;
}

export class OcrService {
  private worker: Worker | null = null;

  async initialize(): Promise<void> {
    if (!this.worker) {
      console.log('Initializing Tesseract worker...');
      this.worker = await createWorker('eng');
      console.log('Tesseract worker initialized');
    }
  }

  async cleanup(): Promise<void> {
    if (this.worker) {
      await this.worker.terminate();
      this.worker = null;
    }
  }

  /**
   * Extract text from image using OCR
   */
  async extractTextFromImage(imagePath: string): Promise<{ text: string; confidence: number }> {
    await this.initialize();

    try {
      // Preprocess image
      const preprocessedPath = imagePath.replace(/\.(jpg|jpeg|png)$/i, '_processed.png');
      await imagePreprocessor.preprocessImage(imagePath, preprocessedPath);

      // Perform OCR
      const { data } = await this.worker!.recognize(preprocessedPath);

      return {
        text: data.text,
        confidence: data.confidence
      };
    } catch (error) {
      console.error('OCR extraction failed:', error);
      throw new Error(`Failed to extract text from image: ${error}`);
    }
  }

  /**
   * Extract text from PDF
   */
  async extractTextFromPdf(pdfPath: string): Promise<{ text: string; confidence: number }> {
    try {
      const dataBuffer = readFileSync(pdfPath);
      const data = await pdfParse(dataBuffer);

      return {
        text: data.text,
        confidence: 95 // PDF text extraction is generally high confidence
      };
    } catch (error) {
      console.error('PDF extraction failed:', error);
      throw new Error(`Failed to extract text from PDF: ${error}`);
    }
  }

  /**
   * Parse extracted text into structured docket data
   */
  parseDocketText(text: string, confidence: number): ExtractedDocketData {
    const lines = text.split('\n').map(line => line.trim()).filter(line => line.length > 0);

    const result: ExtractedDocketData = {
      lineItems: [],
      rawText: text,
      confidence
    };

    // Pattern matching for common docket fields
    const patterns = {
      supplier: /(?:supplier|vendor|from)[\s:]+([^\n]+)/i,
      docketNumber: /(?:docket|delivery|invoice|doc)[\s#:no.]+([A-Z0-9\-]+)/i,
      poReference: /(?:po|purchase\s*order|p\.o\.|order)[\s#:no.]+([A-Z0-9\-]+[ABC]?)/i,
      deliveryDate: /(?:date|delivered|delivery\s*date)[\s:]+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/i,
    };

    // Extract header fields
    for (const line of lines) {
      if (!result.supplierName) {
        const supplierMatch = line.match(patterns.supplier);
        if (supplierMatch) result.supplierName = supplierMatch[1].trim();
      }

      if (!result.docketNumber) {
        const docketMatch = line.match(patterns.docketNumber);
        if (docketMatch) result.docketNumber = docketMatch[1].trim();
      }

      if (!result.poReference) {
        const poMatch = line.match(patterns.poReference);
        if (poMatch) result.poReference = poMatch[1].trim();
      }

      if (!result.deliveryDate) {
        const dateMatch = line.match(patterns.deliveryDate);
        if (dateMatch) result.deliveryDate = dateMatch[1].trim();
      }
    }

    // Extract line items
    // Look for patterns like: "SKU123  Widget Description  5  EA"
    // or: "Product ABC  10  units"
    const lineItemPattern = /^([A-Z0-9\-]+)?\s+(.+?)\s+(\d+(?:\.\d+)?)\s*([A-Z]+)?$/i;

    for (const line of lines) {
      const match = line.match(lineItemPattern);
      if (match) {
        const [, sku, description, qtyStr, unit] = match;
        const quantity = parseFloat(qtyStr);

        if (quantity > 0 && description.length > 2) {
          result.lineItems.push({
            sku: sku?.trim(),
            description: description.trim(),
            quantity,
            unit: unit?.trim()
          });
        }
      }
    }

    // If no structured line items found, try alternative parsing
    if (result.lineItems.length === 0) {
      result.lineItems = this.parseLineItemsFallback(lines);
    }

    return result;
  }

  /**
   * Fallback parser for line items when main pattern fails
   */
  private parseLineItemsFallback(lines: string[]): ExtractedDocketData['lineItems'] {
    const items: ExtractedDocketData['lineItems'] = [];

    // Look for any line with a number that could be quantity
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const numbers = line.match(/\b(\d+(?:\.\d+)?)\b/g);

      if (numbers && numbers.length > 0) {
        // Take last number as quantity (common in dockets)
        const qty = parseFloat(numbers[numbers.length - 1]);

        if (qty > 0 && qty < 10000) { // Reasonable quantity range
          // Use rest of line as description
          const description = line.replace(/\b\d+(?:\.\d+)?\b/g, '').trim();

          if (description.length > 3) {
            items.push({
              description,
              quantity: qty
            });
          }
        }
      }
    }

    return items;
  }

  /**
   * Full pipeline: extract and parse docket from file
   */
  async processDocket(filePath: string): Promise<ExtractedDocketData> {
    const isPdf = /\.pdf$/i.test(filePath);

    const { text, confidence } = isPdf
      ? await this.extractTextFromPdf(filePath)
      : await this.extractTextFromImage(filePath);

    return this.parseDocketText(text, confidence);
  }
}

export const ocrService = new OcrService();
